File: src/semp_workflow/modules/rdp.py
"""RDP module - rdp.add, rdp.delete, rdp.update."""

from __future__ import annotations

import logging

from ..exceptions import SEMPError
from ..models import ActionResult, ResultStatus
from ..semp.client import SempClient
from ..semp.helpers import check_name_length, clean_payload, coerce_bool, enc
from .base import BaseModule

logger = logging.getLogger(__name__)


def _build_rdp_payload(args: dict) -> dict:
    payload = clean_payload(args)
    if "enabled" in payload:
        payload["enabled"] = coerce_bool(payload["enabled"])
    return payload


class RdpAdd(BaseModule):
    description = "Create a REST Delivery Point (RDP). Skipped if the RDP already exists."
    params = {
        "restDeliveryPointName": {"type": "string",  "required": True,  "description": "Name of the REST Delivery Point"},
        "clientProfileName":     {"type": "string",  "required": False, "description": "Client profile to associate with the RDP", "default": "default"},
        "enabled":               {"type": "boolean", "required": False, "description": "Enable the RDP after creation", "default": "true"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        rdp_name = args.get("restDeliveryPointName", "")
        if not rdp_name:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: restDeliveryPointName")

        if err := check_name_length("restDeliveryPointName", rdp_name):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"restDeliveryPoints/{enc(rdp_name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking RDP: {e}")

        if exists:
            return ActionResult(ResultStatus.SKIPPED, f"RDP '{rdp_name}' already exists")

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would create RDP '{rdp_name}'")

        try:
            payload = _build_rdp_payload(args)
            client.create("restDeliveryPoints", payload)
            return ActionResult(ResultStatus.OK, f"RDP '{rdp_name}' created")
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Failed to create RDP '{rdp_name}': {e}")


class RdpDelete(BaseModule):
    description = "Delete a REST Delivery Point. Skipped if the RDP does not exist."
    params = {
        "restDeliveryPointName": {"type": "string", "required": True, "description": "Name of the REST Delivery Point to delete"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        rdp_name = args.get("restDeliveryPointName", "")
        if not rdp_name:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: restDeliveryPointName")

        if err := check_name_length("restDeliveryPointName", rdp_name):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"restDeliveryPoints/{enc(rdp_name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking RDP: {e}")

        if not exists:
            return ActionResult(ResultStatus.SKIPPED, f"RDP '{rdp_name}' does not exist")

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would delete RDP '{rdp_name}'")

        try:
            client.delete(path)
            return ActionResult(ResultStatus.OK, f"RDP '{rdp_name}' deleted")
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Failed to delete RDP '{rdp_name}': {e}")


class RdpUpdate(BaseModule):
    description = "Update attributes of an existing RDP. Fails if the RDP does not exist."
    params = {
        "restDeliveryPointName": {"type": "string",  "required": True,  "description": "Name of the REST Delivery Point to update"},
        "clientProfileName":     {"type": "string",  "required": False, "description": "Client profile to associate with the RDP"},
        "enabled":               {"type": "boolean", "required": False, "description": "Enable or disable the RDP"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        rdp_name = args.get("restDeliveryPointName", "")
        if not rdp_name:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: restDeliveryPointName")

        if err := check_name_length("restDeliveryPointName", rdp_name):
            return ActionResult(ResultStatus.FAILED, err)

        path = f"restDeliveryPoints/{enc(rdp_name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking RDP: {e}")

        if not exists:
            return ActionResult(ResultStatus.FAILED, f"RDP '{rdp_name}' does not exist")

        payload = _build_rdp_payload(args)
        payload.pop("restDeliveryPointName", None)

        if not payload:
            return ActionResult(ResultStatus.SKIPPED, "No fields to update")

        if dry_run:
            return ActionResult(ResultStatus.DRYRUN, f"Would update RDP '{rdp_name}'")

        try:
            client.update(path, payload)
            return ActionResult(ResultStatus.OK, f"RDP '{rdp_name}' updated")
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Failed to update RDP '{rdp_name}': {e}")


MODULES = {
    "rdp.add": RdpAdd,
    "rdp.delete": RdpDelete,
    "rdp.update": RdpUpdate,
}

File: internal/modules/rdp.go
package modules

import (
	"fmt"

	"semp-workflow/internal/models"
	"semp-workflow/internal/semp"
)

func init() {
	register("rdp.add", rdpAdd{})
	register("rdp.delete", rdpDelete{})
	register("rdp.update", rdpUpdate{})
}

// buildRdpPayload cleans args and coerces the enabled flag.
func buildRdpPayload(args map[string]any) map[string]any {
	payload := semp.CleanPayload(args)
	coerceBoolFields(payload, "enabled")
	return payload
}

type rdpAdd struct{}

func (rdpAdd) Description() string {
	return "Create a REST Delivery Point (RDP). Skipped if the RDP already exists."
}

func (rdpAdd) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "restDeliveryPointName", Type: "string", Required: true, Description: "Name of the REST Delivery Point"},
		{Name: "clientProfileName", Type: "string", Description: "Client profile to associate with the RDP", Default: "default"},
		{Name: "enabled", Type: "boolean", Description: "Enable the RDP after creation", Default: "true"},
	}
}

func (rdpAdd) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	rdpName := argStr(args, "restDeliveryPointName", "")
	if rdpName == "" {
		return failed("Missing required arg: restDeliveryPointName")
	}
	if err := semp.CheckNameLength("restDeliveryPointName", rdpName); err != "" {
		return failed(err)
	}

	path := "restDeliveryPoints/" + semp.Enc(rdpName)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking RDP: %s", err))
	}
	if exists {
		return skipped(fmt.Sprintf("RDP '%s' already exists", rdpName))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would create RDP '%s'", rdpName))
	}

	if _, err := c.Create("restDeliveryPoints", buildRdpPayload(args)); err != nil {
		return failed(fmt.Sprintf("Failed to create RDP '%s': %s", rdpName, err))
	}
	return ok(fmt.Sprintf("RDP '%s' created", rdpName))
}

type rdpDelete struct{}

func (rdpDelete) Description() string {
	return "Delete a REST Delivery Point. Skipped if the RDP does not exist."
}

func (rdpDelete) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "restDeliveryPointName", Type: "string", Required: true, Description: "Name of the REST Delivery Point to delete"},
	}
}

func (rdpDelete) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	rdpName := argStr(args, "restDeliveryPointName", "")
	if rdpName == "" {
		return failed("Missing required arg: restDeliveryPointName")
	}
	if err := semp.CheckNameLength("restDeliveryPointName", rdpName); err != "" {
		return failed(err)
	}

	path := "restDeliveryPoints/" + semp.Enc(rdpName)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking RDP: %s", err))
	}
	if !exists {
		return skipped(fmt.Sprintf("RDP '%s' does not exist", rdpName))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would delete RDP '%s'", rdpName))
	}

	if err := c.Delete(path); err != nil {
		return failed(fmt.Sprintf("Failed to delete RDP '%s': %s", rdpName, err))
	}
	return ok(fmt.Sprintf("RDP '%s' deleted", rdpName))
}

type rdpUpdate struct{}

func (rdpUpdate) Description() string {
	return "Update attributes of an existing RDP. Fails if the RDP does not exist."
}

func (rdpUpdate) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "restDeliveryPointName", Type: "string", Required: true, Description: "Name of the REST Delivery Point to update"},
		{Name: "clientProfileName", Type: "string", Description: "Client profile to associate with the RDP"},
		{Name: "enabled", Type: "boolean", Description: "Enable or disable the RDP"},
	}
}

func (rdpUpdate) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	rdpName := argStr(args, "restDeliveryPointName", "")
	if rdpName == "" {
		return failed("Missing required arg: restDeliveryPointName")
	}
	if err := semp.CheckNameLength("restDeliveryPointName", rdpName); err != "" {
		return failed(err)
	}

	path := "restDeliveryPoints/" + semp.Enc(rdpName)

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking RDP: %s", err))
	}
	if !exists {
		return failed(fmt.Sprintf("RDP '%s' does not exist", rdpName))
	}

	payload := buildRdpPayload(args)
	delete(payload, "restDeliveryPointName")

	if len(payload) == 0 {
		return skipped("No fields to update")
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would update RDP '%s'", rdpName))
	}

	if _, err := c.Update(path, payload); err != nil {
		return failed(fmt.Sprintf("Failed to update RDP '%s': %s", rdpName, err))
	}
	return ok(fmt.Sprintf("RDP '%s' updated", rdpName))
}
